examples/swift_package_registry_example/Package.swift

```diff
@@ -0,0 +1,10 @@
+// swift-tools-version: 5.7
+
+import PackageDescription
+
+let package = Package(
+    name: "swift_package_registry_example",
+    dependencies: [
+        .package(id: "apple.swift-collections", exact: "1.1.3"),
```

Comment:
I ran into problems at first with `--replace-scm-with-registry`, because Package.resolved will contain a scoped ID like `apple.swift-collections` but every Package.swift will still reference the unscoped `swift-collections`.

To support that, I made changes to `swiftpkg/internal/deps_indexes.bzl` to remove the scope in `_new_product_index_key`. Not sure that's the ideal fix, but it worked.

Anyway, that's definitely something I would suggest testing out.

Reply:
> To support that, I made changes to swiftpkg/internal/deps_indexes.bzl to remove the scope in _new_product_index_key. Not sure that's the ideal fix, but it worked.

Ooh could you elaborate a bit here, is this an issue with `swift package resolve`, etc or with the way the rules generate something. I dont actually know whats required in a normal SPM project to go from SCM -> Registry, do folks need to delete the Package.resolved and re-resolve to replace the SCM identities?

Reply:
Just an issue with the rules I think.

When you run `swift package resolve --replace-scm-with-registry` the resolver does an inline replacement in all dependencies using the URL. If you run with `--verbose` you'll see messages like this every time it does a replacement of a dependency.

```
info: 'swift-async-algorithms': swizzling 'https://github.com/apple/swift-collections.git' with registry dependency 'apple.swift-collections'.
```

Then Package.resolved gets rewritten with `apple.swift-collections` instead of `swift-collections`. Unfortunately it doesn't record any additional info about the swap. So RSPM is left to deal with packages that still reference a dependency on `swift-collections` but a Package.resolved that doesn't contain it.

I think there's at least a couple options to deal with it:

1. Assume it's OK to ignore the scope during resolution, and index products from `apple.swift-collections` as `swift-collections`. This is easy and worked for me, but it will probably fail if there's ever two products that differ only by scope, with identical product names and package names. I wonder if `swift package resolve` could even handle that.
2. Have RSPM do its own reverse lookup and swizzle dependencies, duping some of the behavior of `swift package resolve`. I'm not sure what that would entail internally.

Reply:
This is the last action item I need before merging this. Any thoughts here @cgrindel @brentleyjones, I may go with option 1 to start unless we have other ideas here.

Reply:
> it will probably fail if there's ever two products that differ only by scope, with identical product names and package names

Is this case when the client specifies the same Swift package twice?

In general, I am fine with option 1.

Reply:
I would prefer option 2 if we can do it. Can we at least look at what SPM is doing for it's "swizzle" and see if we can replicate it?

Reply:
Yeah, I think we'd need to pre-fetch all possible metadata and save it for use within the rules. The rules would then replace URL identities with registry identities.

We'd have a `replace_scm_with_registry` option in the `swift_deps.from_package` call.

Reply:
If that makes sense, let me know. I think we could do this as a follow up as what I have right now works for the basic use-case of "i want to use a registry and dont care if everything is downloaded from the registry"

Reply:
My opinion hasn't changed. I am happy with option 1 for now.

Reply:
Let's go with that then and follow-up as needed

Reply:
> > it will probably fail if there's ever two products that differ only by scope, with identical product names and package names
> 
> Is this case when the client specifies the same Swift package twice?

Haven't tested this, but I think one failure mode here is having two dependencies that independently depend on different packages with the same name *and* a product with the same name. For example:

- the root package depends on package A and package B
- package A depends on `github.com/foo/something`, product named `Common`
- package B depends on `github.com/bar/something`, product named `Common`
- `foo.something` is in the registry

In this case option 1 will resolve both deps to `foo.something`.

But I think this is a rare edge case, and like the idea to follow-up with improvements later.

> I believe the way this works is that SPM makes requests to the registry server to get the metadata for a package which typically returns a github url, from there it can replace the github url with the registry ID.

Almost, but reversed; SwiftPM [queries the registry with each URL](https://github.com/swiftlang/swift-package-manager/blob/main/Documentation/PackageRegistry/Registry.md#endpoint-5) to see if there's an associated registry package. Since we'll already have all the resolved registry IDs in Package.resolved, we could probably create a mapping by querying it in either direction.

Reply:
Apologies for the long pause here, I got time to finish up the implementation and I'm happy with it as a first pass now. I decided to do the same thing SPM does here, e.g. we will look up any available identifiers from the registry given the source control URL of a package.